File: generics_example/src/main/java/com/github/evgzakharov/java/Application2.java
package com.github.evgzakharov.java;

import java.util.List;
import java.util.Map;

public class Application2 {
    public static void main(String[] args) {
        Pair<Long, Long> point1 = new Pair<>(10L, 20L);

        Point point2 = new Point(10L, 10L);

        Map<String, Map<String, List<Pair<Long, Long>>>> someVeryBigMap = null;

        List<Pair<Long, Long>> values = someVeryBigMap.get("first").get("second");

        //List<Point> values2 = someVeryBigMap.get("first").get("second");
    }
}

class Pair<T1, T2> {
    private T1 first;
    private T2 second;

    public Pair(T1 first, T2 second) {
        this.first = first;
        this.second = second;
    }

    public T1 getFirst() {
        return first;
    }

    public T2 getSecond() {
        return second;
    }
}

class Point extends Pair<Long, Long> {
    public Point(Long first, Long second) {
        super(first, second);
    }
}

File: generics_example/src/main/kotlin/com/github/evgzakharov/kotlin/Application2.kt
package com.github.evgzakharov.kotlin

typealias Point = Pair<Long, Long>

typealias MapOfUsersCountryPoints = Map<String, Map<String, List<Point>>>

fun main(args: Array<String>) {

    val point1 = Pair(10L, 10L)

    val point2: Pair<Long, Long> = Pair(10L, 10L)
    val point3: Point = Pair(10L, 10)

    val point4 = Point(10L, 10L)


    val someVeryBigMap: Map<String, Map<String, List<Pair<Long, Long>>>>? = null

    val someVeryBigMap2: Map<String, Map<String, List<Point>>>? = null

    val someVeryBigMap3: MapOfUsersCountryPoints? = null

    val values: List<Point> = someVeryBigMap.geT("first").geT("second")

    val values2: List<Point> = someVeryBigMap2.geT("first").geT("second")

    val values3: List<Point> = someVeryBigMap3.geT("user1").geT("RU")
}



fun <T1, T2> Map<T1,T2>?.geT(key: T1): T2 = this!![key]!!